### PR TITLE
controller: use local less rigid `Selector` interface

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/11480

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: controller/selector.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package controller
+
+import (
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type Selector interface {
+	Matches(labels Labels) bool
+}
```
Comment (kopiczko) [inline controller/selector.go:9]: This new implementation is just leaned out implementation of the [apiextensions Selector type](https://pkg.go.dev/k8s.io/apimachinery@v0.16.6/pkg/labels?tab=doc#Selector).

Now with the leaner interface we can confidently implement more robust conditions. Including OR conditional for labels existence.

**Comment by kopiczko** [comment]
With kubernetes label selector implementation it isn't possible to select an object that has label `A` or label `B`.

More concrete example: it isn't possible to select object having label `cluster` OR `giantswarm.io/cluster`.


This selector `cluster,giantswarm.io/cluster` will selector objects that have both `cluster` AND `giantswarm.io/cluster` labels.

With then new implementation this is possible:

```go
controller.NewSelector(func (labels controller.Labels) bool {
	return labels.Has("cluster") || labels.Has("giantswarm.io/cluster")
})
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,9 +11,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 
 
+### Added
+
+- Add local less rigid `Selector` interface type in `controller.Controller`.
+  The new `Selector` interface is backward compatible with previously used
+  `apiextensions` implementation.
```
Comment (kopiczko) [inline CHANGELOG.md:18]:
```suggestion
  `apiextensions` implementation. #407
```

File: controller/selector.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type Labels = labels.Labels
+
+type Selector interface {
+	Matches(labels Labels) bool
+}
+
```
Comment (xh3b4sd) [inline controller/selector.go:12]:
```suggestion

// NewSelector can be used for custom label matching purposes. The example below shows how 
// one may watch runtime objects having either of the specified labels. Note that there is no native 
// support for the logical OR operator in label selectors, which is why this implementation is useful. 
//
//     controller.NewSelector(func (labels controller.Labels) bool {
//         return labels.Has("cluster") || labels.Has("giantswarm.io/cluster")
//     })
//
```

**Pull request merged**